### DISCUSSION TITLE
How can I conditional transition in invocation?

### DISCUSSION THREAD

**Discussion by baeharam** [statelyai/xstate#1331] [General] [answered]
I know `guards` and `cond` property enables conditional transition. However, I wanna this in invocation which means guarded transition using server response. To achieve this functionality, I have to know two things.

1) What is the order of `actions` and `cond`? Which is first? Does `actions` occur after `cond` or before?
2) When I make specific guard and specify in `guards` property, the type is different from default guard and invocation guard. It means that default guard uses basic event type, but invocation guard uses `DoneInvokeEvent` type. Therefore, I cannot specify guard in `guards` property when I use server response. How can I achieve this?

**Example**

```typescript
const isResultExist = (_: SomeContextType, event: DoneInvokeEvent<string>): boolean => event.data.length > 0;
...
{
  invoke: {
    ... 
   onDone: {
     target: SomeState,
     cond: isResultExist
   }
 }
},
{
  guards: {
    isResultExist // Type Error
  }
}
```

**Comment by Andarist** [comment, marked as answer]
> What is the order of actions and cond? Which is first? Does actions occur after cond or before?

`actions` **won't** be executed if `cond` returns `false`, so naturally `cond` happens first. I'm talking about `{ cond, actions }` situation.

> When I make specific guard and specify in guards property, the type is different from default guard and invocation guard. It means that default guard uses basic event type, but invocation guard uses DoneInvokeEvent type. Therefore, I cannot specify guard in guards property when I use server response. How can I achieve this?

When you put things in `options.guards` they become context-less for us, we can't infer too specific types based on how and where those are used. You can use `as` assertion within your `guards` bodies to assign "correct" type - which is far from ideal but we don't have a better story around this right now. On your place, I would stick to defining them inline if you want to have better type inference.
